pyfixest/estimation/feglm_.py

```diff
@@ -0,0 +1,520 @@
+
+        # _update for inference
+        self._weights = W
+        self._irls_weights = W
+
+        # if only one dim
+        if self._weights.ndim == 1:
+            self._weights = self._weights.reshape((self._N, 1))
+
+        self._u_hat_response = (self._Y.flatten() - self._get_mu(theta=eta)).flatten()
+        self._u_hat_working = (
+            self._u_hat_response
+            if self._method == "feglm-gaussian"
+            else (v_dotdot / W_tilde).flatten()
+        )
+
+        self._scores_response = self._u_hat_response[:, None] * self._X
+        self._scores_working = self._u_hat_working[:, None] * self._X
+
+        self._scores = self._get_score(y=_Y.flatten(), X=_X, mu=mu, eta=eta)
+
+        self._u_hat = self._u_hat_working
+        self._tZX = np.transpose(self._Z) @ self._X
+        self._tZXinv = np.linalg.inv(self._tZX)
+        self._Xbeta = eta
+
+        self._hessian = X_dotdot.T @ X_dotdot
+        self.deviance = deviance
+
+        if _convergence:
+            self._convergence = True
+
+    def _vcov_iid(self):
+        return self._bread
+
+    def _update_v(
+        self, y: np.ndarray, mu: np.ndarray, detadmu: np.ndarray
+    ) -> np.ndarray:
+        "Get (running) dependent variable v for the GLM family."
+        return (y - mu) * detadmu
+
+    def _update_W(self, mu: np.ndarray) -> np.ndarray:
+        "Get (running) weights W for the GLM family."
+        return 1 / (self._update_detadmu(mu=mu) ** 2 * self._get_V(mu=mu))
+        # return (mu * (1 - mu)).reshape(-1, 1)
+
+    def _update_W_tilde(self, W: np.ndarray) -> np.ndarray:
+        "Get W_tilde (formula 3.2)."
+        return np.sqrt(W)
+
+    def _update_v_tilde(
+        self, y: np.ndarray, mu: np.ndarray, W_tilde: np.ndarray, detadmu: np.ndarray
+    ) -> np.ndarray:
+        "Get v_tilde (formula 3.2)."
+        return W_tilde * ((y - mu) * detadmu)
+
+    def _update_X_tilde(self, W_tilde: np.ndarray, X: np.ndarray) -> np.ndarray:
+        "Get X_tilde (formula 3.2)."
+        return W_tilde.reshape(-1, 1) * X
+
+    def _update_beta_diff(
+        self, X_dotdot: np.ndarray, v_dotdot: np.ndarray
+    ) -> np.ndarray:
+        "Get the beta _update difference (formula 3.5) via WLS fit."
+        beta_diff = np.linalg.lstsq(X_dotdot, v_dotdot.reshape(-1, 1), rcond=None)[
+            0
+        ].flatten()
+        return beta_diff
+
+    def _update_eta(
+        self,
+        W_tilde: np.ndarray,
+        v_tilde: np.ndarray,
+        v_dotdot: np.ndarray,
+        X_dotdot: np.ndarray,
+        beta_diff: np.ndarray,
+        eta: np.ndarray,
+    ) -> np.ndarray:
+        "Get the eta _update (formula 4.5)."
+        return eta + X_dotdot @ beta_diff / W_tilde
+
+    def _get_gradient(self, Z: np.ndarray, W: np.ndarray, v: np.ndarray) -> np.ndarray:
+        return Z.T @ W @ v
+
+    def _get_diff(self, deviance: np.ndarray, last: np.ndarray) -> np.ndarray:
+        return np.abs(deviance - last) / (0.1 + np.abs(last))
```

Comment:
why 0.1? It clearly seems to be numerical stability, but could it be smaller (just curious)?

Reply:
I actually don't know - I am using this as it is mimic `fixest`, which itself follows R's glm implementation, which is using this convergence criterion: [link](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/glm.control.html)